spring-webmvc/src/main/java/org/springframework/web/servlet/observability/HttpServletRequestWrapper.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2012-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.web.servlet.observability;
+
+import java.util.Collection;
+import java.util.Collections;
+
+import io.micrometer.core.instrument.transport.http.HttpServerRequest;
+import jakarta.servlet.http.HttpServletRequest;
+
+import org.springframework.web.servlet.HandlerMapping;
+
+/**
+ * An {@link HttpServerRequest} that wraps/delegates to {@link HttpServletRequest}.
+ *
+ * @author Marcin Grzejszczak
+ * @author Jonatan Ivanov
+ *
+ * @since 6.0.0
+ */
+public final class HttpServletRequestWrapper implements HttpServerRequest {
```

Comment:
I wonder if this can be package private? It looks like it's not pluggable. Possible also worth renaming since it's specifically for MVC.

Reply:
Couple of related Boot issues if we're redesigning this:
https://github.com/spring-projects/spring-boot/issues/29303
https://github.com/spring-projects/spring-boot/issues/13064